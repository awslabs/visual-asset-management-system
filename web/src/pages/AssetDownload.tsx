import { useLocation } from "react-router";
import { Storage } from "aws-amplify";
import { FileTree } from "../components/filemanager/FileManager";
import { FileUploadTable, FileUploadTableItem } from "./AssetUpload/FileUploadTable";
import { useReducer, useState } from "react";

async function downloadAllFilesRecursively(tree: FileTree, directoryHandle: any, dispatch: any) {
    for (let subtree of tree.subTree) {
        if (subtree.subTree.length > 0) {
            const subFolderDirectoryHandle = await directoryHandle.getDirectoryHandle(
                subtree.name,
                { create: true }
            );
            await downloadAllFilesRecursively(subtree, subFolderDirectoryHandle, dispatch);
        } else {
            const fileHandle = await directoryHandle.getFileHandle(subtree.name, { create: true });
            const writable = await fileHandle.createWritable();
            const data = await Storage.get(subtree.keyPrefix, {
                download: true,
                progressCallback: (progress: any) => {
                    console.log("Progress callback ", progress);
                    dispatch({
                        type: "UPDATE_PROGRESS",
                        payload: {
                            status: "In Progress",
                            relativePath: subtree.relativePath,
                            progress: progress.loaded,
                            loaded: progress.loaded,
                            total: progress.total,
                        },
                    });
                },
            });
            await writable.write(data.Body);
            await writable.close();
            dispatch({
                type: "UPDATE_STATUS",
                payload: { relativePath: subtree.relativePath, status: "Completed" },
            });
        }
    }
}

async function downloadFolder(tree: FileTree, dispatch: any) {
    //@ts-ignore
    const directoryHandle = await window.showDirectoryPicker();
    await downloadAllFilesRecursively(tree, directoryHandle, dispatch);
}
const convertFileTreeItemsToFileUploadTableItems = (fileTree: FileTree): FileUploadTableItem[] => {
    const allItems: FileUploadTableItem[] = [];
    for (let subtree of fileTree.subTree) {
        if (subtree.subTree.length > 0) {
            allItems.push(...convertFileTreeItemsToFileUploadTableItems(subtree));
        } else {
            allItems.push({
                name: subtree.name,
                index: 0,
                size: 0,
                relativePath: subtree.relativePath,
                status: "Queued",
                progress: 0,
                startedAt: Date.now(),
                loaded: 0,
                total: 0,
            });
        }
    }
    return allItems;
};

const updateIndices = (fileUploadTableItems: FileUploadTableItem[]): FileUploadTableItem[] => {
    let index = 0;
    for (let item of fileUploadTableItems) {
        item.index = index;
        index++;
    }
    return fileUploadTableItems;
};

function assetDownloadReducer(
    state: FileUploadTableItem[],
    action: { type: string; payload: any }
): FileUploadTableItem[] {
    switch (action.type) {
        case "UPDATE_INDICES":
            return updateIndices(state);
        case "UPDATE_PROGRESS":
            return state.map((item) => {
                if (item.relativePath === action.payload.relativePath) {
                    return {
                        ...item,
                        status: action.payload.status,
                        size: action.payload.total,
                        progress: Math.floor((action.payload.loaded / action.payload.total) * 100),
                        loaded: action.payload.loaded,
                        total: action.payload.total,
                    };
                } else {
                    return item;
                }
            });
        case "UPDATE_STATUS":
            return state.map((item) => {
                if (item.relativePath === action.payload.relativePath) {
                    return {
                        ...item,
                        status: action.payload.status,
                    };
                } else {
                    return item;
                }
            });
        default:
            return state;
    }
}
export default function AssetDownloadsPage() {
    const { state } = useLocation();
    const fileTree = state["fileTree"] as FileTree;
    const [resume, setResume] = useState(true);
    const fileUploadTableItems = convertFileTreeItemsToFileUploadTableItems(fileTree);
    const [fileUploadTableItemsState, dispatch] = useReducer(
        assetDownloadReducer,
        fileUploadTableItems
    );
    return (
        <div>
            <h1> Downloading {fileTree.relativePath} </h1>
            <FileUploadTable
                allItems={fileUploadTableItemsState}
                resume={resume}
                onRetry={() => {
                    downloadFolder(fileTree, dispatch).then(() => {
                        setResume(false);
                    });
                }}
                mode={"Download"}
            />
        </div>
    );
}
