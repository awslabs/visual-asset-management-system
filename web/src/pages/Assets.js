/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { useNavigate } from "react-router-dom";
import { fetchAllAssets, fetchDatabaseAssets } from "../services/APIService";
import { AssetListDefinition } from "../components/list/list-definitions/AssetListDefinition";
import ListPage from "./ListPage";
import { useParams } from "react-router";
import Synonyms from "../synonyms";

export default function Assets() {
    const navigate = useNavigate();

    const urlParams = useParams();
    return (
        <ListPage
            singularName={Synonyms.Asset}
            singularNameTitleCase={Synonyms.Asset}
            pluralName={Synonyms.assets}
            pluralNameTitleCase={Synonyms.Assets}
            onCreateCallback={() => {
                if (urlParams.databaseId) {
                    navigate(`/upload/${urlParams.databaseId}`);
                } else {
                    navigate("/upload");
                }
            }}
            listDefinition={AssetListDefinition}
            fetchAllElements={fetchAllAssets}
            fetchElements={fetchDatabaseAssets}
        />
    );
}
