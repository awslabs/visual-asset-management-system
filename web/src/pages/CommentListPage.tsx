/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from "react";

import {
    Box,
    BreadcrumbGroup,
    Button,
    Grid,
    SpaceBetween,
    TextContent,
} from "@cloudscape-design/components";
import { useParams } from "react-router";
import CommentTableList from "../components/list/CommentTableList";
import PropTypes, { object } from "prop-types";
import ListDefinition from "../components/list/list-definitions/types/ListDefinition";
import RelatedTableList from "../components/list/RelatedTableList";
import Synonyms from "../synonyms";

export default function CommentListPage(props: any) {
    const { databaseId } = useParams();
    const {
        singularNameTitleCase,
        pluralName,
        pluralNameTitleCase,
        listDefinition,
        CreateNewElement,
        fetchElements,
        fetchAllElements,
        onCreateCallback,
        isRelatedTable,
        editEnabled,
        onSelection,
        selectedItems,
    } = props;
    const [reload, setReload] = useState<boolean>(true);
    const [loading, setLoading] = useState<boolean>(true);
    const [allItems, setAllItems] = useState<Array<any>>([]);

    const [openNewElement, setOpenNewElement] = useState(false);

    useEffect(() => {
        const getData = async () => {
            setLoading(true);
            let items;
            if (databaseId) {
                items = await fetchElements({ databaseId: databaseId });
            } else {
                items = await fetchAllElements();
            }

            if (items !== false && Array.isArray(items)) {
                setLoading(false);
                setReload(false);
                setAllItems(
                    //@todo fix workflow delete return
                    items.filter((item) => item.databaseId.indexOf("#deleted") === -1)
                );
            }
        };
        if (reload) {
            getData();
        }
    }, [reload, databaseId, fetchAllElements, fetchElements]);

    const handleOpenNewElement = () => {
        if (onCreateCallback) onCreateCallback();
        else if (CreateNewElement) setOpenNewElement(true);
    };

    return (
        <>
            <Box padding={{ top: databaseId ? "s" : "m", horizontal: "l" }}>
                {databaseId && (
                    <BreadcrumbGroup
                        items={[
                            { text: Synonyms.Databases, href: "/databases/" },
                            {
                                text: databaseId,
                                href: `/databases/${databaseId}/${pluralName}/`,
                            },
                        ]}
                        ariaLabel="Breadcrumbs"
                    />
                )}
                <Grid gridDefinition={[{ colspan: { default: 6 } }]}>
                    <div>
                        <TextContent>
                            <h1>
                                {pluralNameTitleCase}
                                {databaseId && ` for ${databaseId}`}
                            </h1>
                        </TextContent>
                    </div>
                </Grid>
                <Grid gridDefinition={[{ colspan: { default: 12 } }]}>
                    {isRelatedTable && (
                        <RelatedTableList
                            allItems={allItems}
                            loading={loading}
                            listDefinition={listDefinition}
                            databaseId={databaseId}
                            setReload={setReload}
                        />
                    )}
                    <CommentTableList
                        allItems={allItems}
                        loading={loading}
                        listDefinition={listDefinition}
                        databaseId={databaseId}
                        editEnabled={editEnabled}
                        setReload={setReload}
                        onSelection={onSelection}
                        selectedItems={selectedItems}
                        UpdateSelectedElement={CreateNewElement}
                        createNewElement={
                            (CreateNewElement || onCreateCallback) && (
                                <div style={{ float: "right" }}>
                                    <SpaceBetween direction={"horizontal"} size={"m"}>
                                        <Button onClick={handleOpenNewElement} variant="primary">
                                            Create {singularNameTitleCase}
                                        </Button>
                                    </SpaceBetween>
                                </div>
                            )
                        }
                    />
                </Grid>
            </Box>
            {CreateNewElement && (
                <CreateNewElement
                    open={openNewElement}
                    setOpen={setOpenNewElement}
                    setReload={setReload}
                    databaseId={databaseId}
                />
            )}
        </>
    );
}

CommentListPage.propTypes = {
    singularName: PropTypes.string.isRequired,
    singularNameTitleCase: PropTypes.string.isRequired,
    pluralName: PropTypes.string.isRequired,
    pluralNameTitleCase: PropTypes.string.isRequired,
    listDefinition: PropTypes.instanceOf(ListDefinition).isRequired,
    CreateNewElement: PropTypes.func,
    fetchElements: PropTypes.func.isRequired,
    fetchAllElements: PropTypes.func,
    onCreateCallback: PropTypes.func,
    onSelection: PropTypes.func,
    selectedItems: PropTypes.arrayOf(object),
    isRelatedTable: PropTypes.bool,
    editEnabled: PropTypes.bool,
};
