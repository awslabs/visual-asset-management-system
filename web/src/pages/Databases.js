/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from "react";
import { fetchAllDatabases } from "../services/APIService";
import CreateDatabase from "../components/createupdate/CreateDatabase";
import { DatabaseListDefinition } from "../components/list/list-definitions/DatabaseListDefinition";
import ListPage from "./ListPage";

import { anyRoleOf } from "../FedAuth/roles";

export default function Databases() {
    const [createDatabase, setCreateDatabase] = useState(false);
    useEffect(() => {
        anyRoleOf(["create-database", "super-admin"]).then((result) => {
            setCreateDatabase(result);
        });
    }, []);
    return (
        <ListPage
            singularName={"database"}
            singularNameTitleCase={"Database"}
            pluralName={"databases"}
            pluralNameTitleCase={"Databases"}
            listDefinition={DatabaseListDefinition}
            CreateNewElement={(createDatabase && CreateDatabase) || undefined}
            fetchAllElements={fetchAllDatabases}
            fetchElements={fetchAllDatabases}
            editEnabled={true}
        />
    );
}
