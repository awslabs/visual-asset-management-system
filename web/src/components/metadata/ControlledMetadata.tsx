/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { useEffect, useState } from "react";
import { SchemaContextData } from "../../pages/MetadataSchema";
import { API, Storage } from "aws-amplify";
import { Container, Grid, Header } from "@cloudscape-design/components";
import React from "react";
import { EditComp } from "./EditComp";
import { HandleControlData, handleCSVControlData as originHandler } from "./CSVControlData";
import MetadataTable, { MetadataApi, put } from "../single/Metadata";

export interface Metadata {
    [k: string]: string;
}

interface ControlledMetadataProps {
    assetId: string;
    databaseId: string;
    initialState?: Metadata;
    store?: (databaseId: string, assetId: string, record: Metadata) => Promise<any>;
    apiget?: (apiName: string, path: string, init: any) => Promise<any>;
    storageget?: (key: string) => Promise<any>;
    handleCSVControlData?: HandleControlData;
}

export interface TableRow {
    idx: number;
    name: string;
    value: string;
    inlineValues: string[];
    type: string;
    dependsOn: string[];
}

export default function ControlledMetadata({
    databaseId,
    assetId,
    initialState,
    apiget = API.get.bind(API),
    storageget = Storage.get.bind(Storage),
    handleCSVControlData = originHandler,
    store = put,
}: ControlledMetadataProps) {
    const [schema, setSchema] = useState<SchemaContextData | null>(null);
    const [controlledLists, setControlledLists] = useState<any | null>(null);
    const [rawControlData, setRawControlData] = useState<any>([]);
    const [metadata, setMetadata] = useState<Metadata | null>();
    const [items, setItems] = useState<TableRow[]>([]);

    // console.log("schema", schema);
    // console.log("items", items);
    // console.log("controlledLists", controlledLists);

    const metaToTableRow = (meta: Metadata, schema: SchemaContextData) => {
        schema.schemas.sort((a, b) => {
            if (a.sequenceNumber === undefined) {
                return 1;
            }
            if (b.sequenceNumber === undefined) {
                return -1;
            }
            return a.sequenceNumber - b.sequenceNumber;
        });

        return schema.schemas
            .map((x) => x.field)
            .map((key, idx): TableRow => {
                return {
                    idx: idx,
                    name: key,
                    value: meta[key],
                    dependsOn: schema.schemas.find((x) => x.field === key)?.dependsOn || [],
                    inlineValues:
                        schema.schemas
                            .find((x) => x.field === key)
                            ?.inlineControlledListValues?.split(",")
                            .map((x) => x.trim()) || [],
                    type: schema.schemas.find((x) => x.field === key)?.dataType || "string",
                };
            });
    };

    const tableRowToMeta = (rows: TableRow[]): Metadata => {
        const result: Metadata = {};
        rows.forEach((row) => {
            if (row && row.name && row.value) {
                result[row.name] = row.value;
            }
        });
        return result;
    };

    // given the databaseId, retrieve the metadata schema for the database

    useEffect(() => {
        if (schema !== null) {
            return;
        }

        if (initialState === undefined) {
            apiget("api", `metadata/${databaseId}/${assetId}`, {}).then(
                ({ metadata: start }: MetadataApi) => {
                    apiget("api", `metadataschema/${databaseId}`, {}).then(
                        (data: SchemaContextData) => {
                            setSchema(data);
                            if (data.schemas.length > 0) {
                                const meta = data.schemas.reduce((acc, x) => {
                                    acc[x.field] = start[x.field] || "";
                                    return acc;
                                }, start);
                                console.log("metadata in init", meta);
                                setMetadata(meta);
                                setItems(metaToTableRow(meta, data));
                            }
                        }
                    );
                }
            );
        } else {
            apiget("api", `metadataschema/${databaseId}`, {}).then((data: SchemaContextData) => {
                setSchema(data);
                if (data.schemas.length > 0) {
                    const start: Metadata = initialState || {};
                    const meta = data.schemas.reduce((acc, x) => {
                        acc[x.field] = start[x.field] || "";
                        return acc;
                    }, start);
                    setMetadata(meta);
                    setItems(metaToTableRow(meta, data));
                }
            });
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [databaseId, assetId, initialState]);

    useEffect(() => {
        if (schema === null || schema.schemas.length === 0) return;
        if (controlledLists !== null) return;
        storageget(`metadataschema/${databaseId}/controlledlist.csv`).then((url: string) => {
            handleCSVControlData(url, schema, setControlledLists, setRawControlData);
        });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [databaseId, schema]);

    if (schema && schema.schemas.length === 0) {
        return (
            <MetadataTable
                assetId={assetId || ""}
                databaseId={databaseId || ""}
                initialState={initialState}
                store={store}
                data-testid="metadata-table"
            />
        );
    }
    if (!metadata || !controlledLists || !schema) {
        return <p>Loading...</p>;
    }

    return (
        <React.Fragment>
            <Container
                header={
                    <Header variant="h2" description="Metadata">
                        Metadata
                    </Header>
                }
            >
                {items.map((row) => {
                    return (
                        <Grid gridDefinition={[{ colspan: 3 }, { colspan: 6 }]} key={row.idx}>
                            <div>{row.name}</div>
                            <div>
                                <EditComp
                                    controlledLists={controlledLists}
                                    metadata={metadata}
                                    item={row}
                                    currentValue={row.value}
                                    schema={schema}
                                    controlData={rawControlData}
                                    setValue={(value) => {
                                        if (value !== undefined) {
                                            console.log("items", items);
                                            const next: TableRow[] = [...items];
                                            next[row.idx].value = value;

                                            const resetDeps = (name: string) => {
                                                for (const item of next) {
                                                    if (item.dependsOn.includes(name)) {
                                                        item.value = "";
                                                        resetDeps(item.name);
                                                    }
                                                }
                                            };

                                            resetDeps(row.name);

                                            setItems(next);
                                            setMetadata(tableRowToMeta(next));
                                            if (store)
                                                store(databaseId, assetId, tableRowToMeta(next));
                                        } else {
                                            console.log("undefined value", row);
                                        }
                                    }}
                                />
                            </div>
                        </Grid>
                    );
                })}
            </Container>
        </React.Fragment>
    );
}
