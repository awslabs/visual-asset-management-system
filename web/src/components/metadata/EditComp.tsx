/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { SchemaContextData } from "../../pages/MetadataSchema";
import { Checkbox, DatePicker, Input, Select, Textarea } from "@cloudscape-design/components";
import { MapLocationSelectorModal2 } from "./MapLocationSelector";
import { TableRow, Metadata } from "./ControlledMetadata";

export interface EditCompProps {
    item: TableRow;
    schema: SchemaContextData;
    controlledLists: any;
    controlData: any;
    setValue: (row: string | undefined) => void;
    currentValue: string;
    metadata: Metadata;
}
export function EditComp({
    item,
    setValue,
    controlledLists,
    currentValue,
    metadata,
    schema,
    controlData,
}: EditCompProps) {
    const disabled = !item.dependsOn.every((x: string) => metadata[x] && metadata[x] !== "");

    if (item.type === "string") {
        return (
            <Input
                value={item.value}
                disabled={disabled}
                onChange={(event) => {
                    setValue(event.detail.value);
                }}
            />
        );
    }

    if (item.type === "textarea") {
        return (
            <Textarea
                disabled={disabled}
                onChange={({ detail }) => setValue(detail.value)}
                value={item.value}
            />
        );
    }

    if (item.type === "inline-controlled-list") {
        const options = item.inlineValues.map((label: string) => ({
            label,
            value: label,
        }));
        let selectedOption = {
            label: currentValue,
            value: currentValue,
        };
        if (options.length === 1 && currentValue !== options[0].value) {
            selectedOption = options[0];
            setValue(selectedOption.value);
        }
        return (
            <Select
                options={options}
                selectedOption={selectedOption}
                disabled={disabled || options.length === 1}
                expandToViewport
                filteringType="auto"
                onChange={(e) => setValue(e.detail.selectedOption.value)}
            />
        );
    }

    if (item.type === "controlled-list" && item.dependsOn.length === 0) {
        const options = controlledLists[item.name].data.map((x: any) => ({
            label: x[item.name],
            value: x[item.name],
        }));
        let selectedOption = {
            label: currentValue,
            value: currentValue,
        };
        if (options.length === 1 && currentValue !== options[0].value) {
            selectedOption = options[0];
            setValue(selectedOption.value);
        }
        return (
            <Select
                options={options}
                selectedOption={selectedOption}
                expandToViewport
                disabled={disabled}
                filteringType="auto"
                onChange={(e) => setValue(e.detail.selectedOption.value)}
            />
        );
    }
    if (item.type === "controlled-list" && item.dependsOn.length > 0) {
        const options = controlledLists[item.name].data
            .filter((x: any) => {
                return item.dependsOn.every((y: string) => {
                    return x[y] === metadata[y];
                });
            })
            .map((x: any) => ({
                label: x[item.name],
                value: x[item.name],
            }));
        let selectedOption = {
            label: currentValue,
            value: currentValue,
        };
        if (options.length === 1 && currentValue !== options[0].value) {
            selectedOption = options[0];
            setValue(selectedOption.value);
        }

        return (
            <Select
                options={options}
                selectedOption={selectedOption}
                disabled={disabled || options.length === 1}
                expandToViewport
                filteringType="auto"
                onChange={(e) => setValue(e.detail.selectedOption.value)}
            />
        );
    }
    if (item.type === "location") {
        let currentValueInit = {
            loc: [-95.37019986475366, 29.767650706163337], // Houston
            zoom: 5,
        };

        if (!currentValue) {
            const schemaItem = schema.schemas.find((x) => x.field === item.name);
            if (schemaItem && schemaItem.dependsOn) {
                const controlDataItem = controlData.find((x: any) =>
                    schemaItem.dependsOn.every((y: string) => x[y] === metadata[y])
                );
                if (controlDataItem) {
                    currentValueInit = {
                        loc: [
                            controlDataItem[schemaItem.longitudeField!],
                            controlDataItem[schemaItem.latitudeField!],
                        ],
                        zoom: controlDataItem[schemaItem.zoomLevelField!],
                    };
                }
            }
        }

        return (
            <MapLocationSelectorModal2
                json={currentValue ? currentValue : JSON.stringify(currentValueInit)}
                setJson={(json) => {
                    setValue(json);
                }}
                disabled={disabled}
            />
        );
    }

    if (item.type === "number") {
        return (
            <Input
                value={item.value}
                inputMode="numeric"
                disabled={disabled}
                onChange={(event) => {
                    setValue(event.detail.value);
                }}
            />
        );
    }

    if (item.type === "boolean") {
        // checkbox
        return (
            <Checkbox
                checked={item.value === "true"}
                disabled={disabled}
                onChange={(e) => {
                    setValue(e.detail.checked ? "true" : "false");
                }}
            />
        );
    }

    if (item.type === "date") {
        return (
            <DatePicker
                onChange={({ detail }) => setValue(detail.value)}
                value={item.value}
                disabled={disabled}
                openCalendarAriaLabel={(selectedDate) =>
                    "Choose date" + (selectedDate ? `, selected date is ${selectedDate}` : "")
                }
                nextMonthAriaLabel="Next month"
                placeholder="YYYY/MM/DD"
                previousMonthAriaLabel="Previous month"
                todayAriaLabel="Today"
            />
        );
    }

    return null;
}
