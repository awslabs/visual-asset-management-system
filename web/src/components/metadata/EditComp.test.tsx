/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { render } from "@testing-library/react";
import { SchemaContextData } from "../../pages/MetadataSchema";
import { EditComp } from "./EditComp";
import { Metadata, TableRow } from "./ControlledMetadata";
import createWrapper from "@cloudscape-design/components/test-utils/dom";

describe("EditComp", () => {
    it("renders string fields", () => {
        const item: TableRow = {
            idx: 0,
            name: "field1",
            value: "value1",
            type: "string",
            dependsOn: [],
            inlineValues: [],
        };
        const schema: SchemaContextData = {
            databaseId: "dbId",
            schemas: [],
        };

        const metadata: Metadata = {};

        const view = render(
            <EditComp
                item={item}
                schema={schema}
                controlledLists={undefined}
                controlData={undefined}
                setValue={function (row: string | undefined): void {
                    throw new Error("Function not implemented.");
                }}
                currentValue={""}
                metadata={metadata}
            />
        );

        expect(view).toMatchSnapshot();
    });

    it("renders textarea fields", () => {
        const item: TableRow = {
            idx: 0,
            name: "field1",
            value: "value1",
            type: "textarea",
            dependsOn: [],
            inlineValues: [],
        };
        const schema: SchemaContextData = {
            databaseId: "dbId",
            schemas: [],
        };

        const metadata: Metadata = {};

        const view = render(
            <EditComp
                item={item}
                schema={schema}
                controlledLists={undefined}
                controlData={undefined}
                setValue={function (row: string | undefined): void {
                    throw new Error("Function not implemented.");
                }}
                currentValue={""}
                metadata={metadata}
            />
        );

        expect(view).toMatchSnapshot();
    });

    it("renders inline controlled lists", () => {
        const item: TableRow = {
            idx: 0,
            name: "field1",
            value: "value1",
            type: "inline-controlled-list",
            dependsOn: [],
            inlineValues: ["opt1", "opt2"],
        };
        const schema: SchemaContextData = {
            databaseId: "dbId",
            schemas: [],
        };

        const metadata: Metadata = {};

        const view = render(
            <EditComp
                item={item}
                schema={schema}
                controlledLists={undefined}
                controlData={undefined}
                setValue={function (row: string | undefined): void {
                    throw new Error("Function not implemented.");
                }}
                currentValue={""}
                metadata={metadata}
            />
        );

        const wrapper = createWrapper(view.container);
        expect(wrapper.findSelect()).toBeTruthy();
        expect(wrapper.findSelect()?.findDropdown()).toBeTruthy();
    });

    it("renders controlled lists", () => {
        const item: TableRow = {
            idx: 0,
            name: "field1",
            value: "value1",
            type: "controlled-list",
            dependsOn: [],
            inlineValues: [],
        };
        const schema: SchemaContextData = {
            databaseId: "dbId",
            schemas: [],
        };

        const controlledLists = {
            field1: {
                data: ["v1", "v2"],
            },
        };

        const metadata: Metadata = {};

        render(
            <EditComp
                item={item}
                schema={schema}
                controlledLists={controlledLists}
                controlData={undefined}
                setValue={function (row: string | undefined): void {
                    throw new Error("Function not implemented.");
                }}
                currentValue={""}
                metadata={metadata}
            />
        );

        const wrapper = createWrapper();
        expect(wrapper.findSelect()).toBeTruthy();
        expect(wrapper.findSelect()?.findDropdown()).toBeTruthy();
    });

    it("renders controlled lists with deps", () => {
        const item: TableRow = {
            idx: 0,
            name: "field1",
            value: "value1",
            type: "controlled-list",
            dependsOn: ["field2"],
            inlineValues: [],
        };
        const schema: SchemaContextData = {
            databaseId: "dbId",
            schemas: [],
        };

        const controlledLists = {
            field1: {
                data: ["v1", "v2"],
            },
        };

        const metadata: Metadata = {};

        render(
            <EditComp
                item={item}
                schema={schema}
                controlledLists={controlledLists}
                controlData={undefined}
                setValue={function (row: string | undefined): void {
                    throw new Error("Function not implemented.");
                }}
                currentValue={""}
                metadata={metadata}
            />
        );

        const wrapper = createWrapper();
        expect(wrapper.findSelect()).toBeTruthy();
        expect(wrapper.findSelect()?.findDropdown()).toBeTruthy();
    });

    it("renders a date input", () => {
        const item: TableRow = {
            idx: 0,
            name: "field1",
            value: "2023/01/01",
            type: "date",
            dependsOn: [],
            inlineValues: [],
        };
        const schema: SchemaContextData = {
            databaseId: "dbId",
            schemas: [],
        };

        const metadata: Metadata = {};

        const view = render(
            <EditComp
                item={item}
                schema={schema}
                controlledLists={undefined}
                controlData={undefined}
                setValue={function (row: string | undefined): void {
                    throw new Error("Function not implemented.");
                }}
                currentValue={""}
                metadata={metadata}
            />
        );

        const wrapper = createWrapper(view.container);
        expect(wrapper.findDatePicker()).toBeTruthy();
    });
});
