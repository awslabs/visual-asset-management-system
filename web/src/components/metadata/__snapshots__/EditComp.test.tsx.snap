// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditComp renders string fields 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="awsui_root_2rhyz_1mdn9_93 awsui_input-container_2rhyz_1mdn9_220"
      >
        <input
          autocomplete="on"
          class="awsui_input_2rhyz_1mdn9_97"
          type="text"
          value="value1"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="awsui_root_2rhyz_1mdn9_93 awsui_input-container_2rhyz_1mdn9_220"
    >
      <input
        autocomplete="on"
        class="awsui_input_2rhyz_1mdn9_97"
        type="text"
        value="value1"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`EditComp renders textarea fields 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="awsui_root_qk1j1_14w5r_93"
      >
        <textarea
          autocomplete="on"
          class="awsui_textarea_qk1j1_14w5r_97"
          rows="3"
        >
          value1
        </textarea>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="awsui_root_qk1j1_14w5r_93"
    >
      <textarea
        autocomplete="on"
        class="awsui_textarea_qk1j1_14w5r_97"
        rows="3"
      >
        value1
      </textarea>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
